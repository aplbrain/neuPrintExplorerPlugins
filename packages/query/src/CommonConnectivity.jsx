/*
 * Queries common inputs/outputs given list of bodyIds
 */
import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import TextField from '@material-ui/core/TextField';

import NeuronFilter from './shared/NeuronFilter';

const styles = theme => ({
  formControl: {
    margin: theme.spacing.unit
  },
  chips: {
    display: 'flex',
    flexWrap: 'wrap'
  }
});

const pluginName = 'CommonConnectivity';
const pluginAbbrev = 'cc';

const groupBy = (inputJson, key) =>
  inputJson.reduce((accumulator, currentValue) => {
    // name of the common input/output
    const { name, type } = currentValue;
    // first element of the keys array is X_weight where X is the body id of a queried neuron
    let weights = Object.keys(currentValue)[0];
    // in case order of keys changes check that this is true and if not find the correct key
    if (!weights.endsWith('weight')) {
      for (let i = 1; i < Object.keys(currentValue).length; i += 1) {
        if (Object.keys(currentValue)[i].endsWith('weight')) {
          weights = Object.keys(currentValue)[i];
          break;
        }
      }
    }
    (accumulator[currentValue[key]] = accumulator[currentValue[key]] || {})[weights] =
      currentValue[weights];
    accumulator[currentValue[key]].name = name;
    accumulator[currentValue[key]].type = type;
    return accumulator;
  }, {});

class CommonConnectivity extends React.Component {
  static get details() {
    return {
      name: pluginName,
      displayName: 'Common connectivity',
      abbr: pluginAbbrev,
      description:
        'Finds common inputs/outputs for a group of bodies and weights of their connections to these inputs/outputs.',
      visType: 'SimpleTable'
    };
  }

  static fetchParameters() {
    return {
      queryString: '/npexplorer/commonconnectivity'
    };
  }

  static processDownload(response) {
    const { find_inputs, neuron_ids, neuron_names } = response.params.pm;

    const queryKey = find_inputs ? 'input' : 'output';
    const connectionArray = response.result.data[0][0];

    const columnHeaders = [
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} ID`,
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} Name`,
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} Type`
    ];

    const groupedByInputOrOutputId = groupBy(connectionArray, queryKey);

    let selectedNeurons = [];
    if (neuron_ids.length > 0) {
      selectedNeurons = neuron_ids;
    } else {
      selectedNeurons = neuron_names;
    }

    const selectedWeightHeadings = selectedNeurons.map(neuron => `${neuron}_weight`);
    selectedWeightHeadings.forEach(neuronWeightHeading => {
      columnHeaders.push(neuronWeightHeading);
    });

    const data = [];
    Object.keys(groupedByInputOrOutputId).forEach(inputOrOutput => {
      const singleRow = [parseInt(inputOrOutput, 10), groupedByInputOrOutputId[inputOrOutput].name, groupedByInputOrOutputId[inputOrOutput].type];
      selectedWeightHeadings.forEach(selectedWeightHeading => {
        const selectedWeightValue =
          groupedByInputOrOutputId[inputOrOutput][selectedWeightHeading] || 0;
        singleRow.push(parseInt(selectedWeightValue, 10));
      });
      data.push(singleRow);
    });

    return [columnHeaders, data.join('\n')].join('\n');

  }

  static processResults(query, apiResponse) {
    const { pm: parameters } = query;

    const queryKey = parameters.find_inputs ? 'input' : 'output';
    const connectionArray = apiResponse.data[0][0];

    const columns = [
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} ID`,
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} Name`,
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} Type`
    ];

    const groupedByInputOrOutputId = groupBy(connectionArray, queryKey);

    let selectedNeurons = [];
    if (parameters.neuron_ids.length > 0) {
      selectedNeurons = parameters.neuron_ids;
    } else {
      selectedNeurons = parameters.neuron_names;
    }

    const selectedWeightHeadings = selectedNeurons.map(neuron => `${neuron}_weight`);
    selectedWeightHeadings.forEach(neuronWeightHeading => {
      columns.push(neuronWeightHeading);
    });

    const data = [];
    Object.keys(groupedByInputOrOutputId).forEach(inputOrOutput => {
      const singleRow = [parseInt(inputOrOutput, 10), groupedByInputOrOutputId[inputOrOutput].name, groupedByInputOrOutputId[inputOrOutput].type];
      selectedWeightHeadings.forEach(selectedWeightHeading => {
        const selectedWeightValue =
          groupedByInputOrOutputId[inputOrOutput][selectedWeightHeading] || 0;
        singleRow.push(parseInt(selectedWeightValue, 10));
      });
      data.push(singleRow);
    });

    return {
      columns,
      data,
      debug: apiResponse.debug,
      title: `Common ${queryKey}s for ${selectedNeurons} in ${parameters.dataset}`
    };
  }

  constructor(props) {
    super(props);
    this.state = {
      limitNeurons: true,
      statusFilters: [],
      preThreshold: 0,
      postThreshold: 0,
      bodyIds: '',
      typeValue: 'input'
    };
  }

  processRequest = () => {
    const { dataSet, submit, actions } = this.props;
    const { limitNeurons, preThreshold, postThreshold, statusFilters } = this.state;
    const { bodyIds, typeValue } = this.state;

    const parameters = {
      dataset: dataSet,
      statuses: statusFilters,
      find_inputs: typeValue !== 'output',
      neuron_ids: bodyIds === '' ? [] : bodyIds.split(',').map(Number),
      all_segments: !limitNeurons
    };

    if (parameters.neuron_ids.length > 100) {
      actions.metaInfoError(`You entered ${parameters.neuron_ids.length} Neuron IDs. Please limit the list to 100 or less`);
      return
    }

    if (preThreshold > 0) {
      parameters.pre_threshold = preThreshold;
    }

    if (postThreshold > 0) {
      parameters.post_threshold = postThreshold;
    }

    const query = {
      dataSet,
      plugin: pluginName,
      pluginCode: pluginAbbrev,
      parameters
    };

    submit(query);
  };

  loadNeuronFilters = params => {
    this.setState({
      limitNeurons: params.limitNeurons,
      statusFilters: params.statusFilters,
      preThreshold: parseInt(params.preThreshold, 10),
      postThreshold: parseInt(params.postThreshold, 10)
    });
  };

  addNeuronBodyIds = event => {
    this.setState({
      bodyIds: event.target.value
    });
  };

  setInputOrOutput = event => {
    this.setState({ typeValue: event.target.value });
  };

  catchReturn = event => {
    // submit request if user presses enter
    if (event.keyCode === 13) {
      event.preventDefault();
      this.processRequest();
    }
  };

  render() {
    const { classes, dataSet, actions, neoServerSettings } = this.props;
    const { bodyIds, typeValue } = this.state;

    return (
      <div>
        <FormControl fullWidth className={classes.formControl}>
          <TextField
            label="Neuron IDs"
            multiline
            fullWidth
            rows={1}
            value={bodyIds}
            name="bodyIds"
            rowsMax={4}
            helperText="Separate IDs with commas. Max 100"
            onChange={this.addNeuronBodyIds}
            onKeyDown={this.catchReturn}
          />
        </FormControl>
        <RadioGroup
          aria-label="Type Of Connections"
          name="type"
          value={typeValue}
          onChange={this.setInputOrOutput}
        >
          <FormControlLabel value="input" control={<Radio color="primary" />} label="Inputs" />
          <FormControlLabel value="output" control={<Radio color="primary" />} label="Outputs" />
        </RadioGroup>
        <NeuronFilter
          callback={this.loadNeuronFilters}
          datasetstr={dataSet}
          actions={actions}
          neoServer={neoServerSettings.get('neoServer')}
        />
        <Button variant="contained" color="primary" onClick={this.processRequest}>
          Submit
        </Button>
      </div>
    );
  }
}

CommonConnectivity.propTypes = {
  dataSet: PropTypes.string.isRequired,
  actions: PropTypes.object.isRequired,
  submit: PropTypes.func.isRequired,
  neoServerSettings: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired
};

export default withStyles(styles, { withTheme: true })(CommonConnectivity);
