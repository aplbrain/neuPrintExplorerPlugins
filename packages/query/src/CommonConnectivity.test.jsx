import { MemoryRouter } from 'react-router-dom';
import CommonConnectivity from './CommonConnectivity';

const styles = {};
const { actions, React, enzyme, renderer, submit } = global;

const neoServerSettings = {
  get: () => 'http://example.com'
};

const raw = (
  <MemoryRouter>
    <CommonConnectivity
      actions={actions}
      submit={submit}
      dataSet="mb6"
      history={{ push: jest.fn() }}
      classes={styles}
      neoServerSettings={neoServerSettings}
    />
  </MemoryRouter>
);

function providedRenderedComponent() {
  const wrapper = enzyme.mount(raw);
  // get through the styles and router components that wrap the plugin.
  const rendered = wrapper.children().children();
  return rendered;
}

describe('Common Connectivity Plugin', () => {
  beforeEach(() => {
    submit.mockClear();
  });

  describe('has required functions', () => {
    test('name', () => {
      expect(CommonConnectivity.details.name).toBeTruthy();
    });
    test('description', () => {
      expect(CommonConnectivity.details.description).toBeTruthy();
    });
  });

  describe('renders correct defaults', () => {
    const rendered = providedRenderedComponent();

    it('should render', () => {
      const component = renderer.create(raw);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('bodyIds should be empty', () => {
      expect(
        rendered
          .find('textarea')
          .at(2)
          .props().value
      ).toEqual('');
    });

    test('typeValue should be inputs', () => {
      const inputType = rendered.find('input[name="type"]').at(0);
      expect(inputType.props().value).toEqual('input');
      expect(inputType.props().checked).toBe(true);

      const outputType = rendered.find('input[name="type"]').at(1);
      expect(outputType.props().value).toEqual('output');
      expect(outputType.props().checked).toBe(false);
    });
  });

  describe('submits correctly', () => {
    const rendered = providedRenderedComponent();
    test('submit button pressed', () => {
      rendered.find('Button').simulate('click');
      expect(submit).toHaveBeenCalledTimes(1);
    });

    test('enter key pressed in text field', () => {
      rendered
        .find('TextField')
        .at(0)
        .simulate('keyDown', { keyCode: 13 });
      expect(submit).toHaveBeenCalledTimes(1);
    });
  });
});
