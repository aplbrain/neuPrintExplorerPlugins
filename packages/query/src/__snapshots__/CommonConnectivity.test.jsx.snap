// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common Connectivity Plugin renders correct defaults should render 1`] = `
<div>
  <div
    className="MuiFormControl-root-3 MuiFormControl-fullWidth-6 CommonConnectivity-formControl-1"
  >
    <div
      className="MuiFormControl-root-3 MuiFormControl-fullWidth-6"
      onKeyDown={[Function]}
    >
      <label
        className="MuiFormLabel-root-18 MuiInputLabel-root-7 MuiInputLabel-formControl-12 MuiInputLabel-animated-15"
        data-shrink={false}
      >
        Neuron IDs
      </label>
      <div
        className="MuiInputBase-root-38 MuiInput-root-25 MuiInput-underline-29 MuiInputBase-fullWidth-47 MuiInput-fullWidth-32 MuiInputBase-formControl-39 MuiInput-formControl-26 MuiInputBase-multiline-46 MuiInput-multiline-31"
        onClick={[Function]}
      >
        <div
          className="MuiPrivateTextarea-root-55"
        >
          <textarea
            aria-hidden="true"
            className="MuiPrivateTextarea-textarea-56 MuiPrivateTextarea-shadow-57"
            readOnly={true}
            rows="1"
            tabIndex={-1}
            value=""
          />
          <textarea
            aria-hidden="true"
            className="MuiPrivateTextarea-textarea-56 MuiPrivateTextarea-shadow-57"
            readOnly={true}
            rows={1}
            tabIndex={-1}
            value=""
          />
          <textarea
            aria-invalid={false}
            className="MuiPrivateTextarea-textarea-56 MuiInputBase-input-48 MuiInput-input-33 MuiInputBase-inputMultiline-50 MuiInput-inputMultiline-35"
            disabled={false}
            name="bodyIds"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            rows={1}
            style={
              Object {
                "height": 19,
              }
            }
            value=""
          />
        </div>
      </div>
      <p
        className="MuiFormHelperText-root-58"
      >
        Separate IDs with commas. Max 100
      </p>
    </div>
  </div>
  <div
    aria-label="Type Of Connections"
    className="MuiFormGroup-root-66"
    role="radiogroup"
  >
    <label
      className="MuiFormControlLabel-root-68"
    >
      <span
        className="MuiButtonBase-root-89 MuiIconButton-root-83 MuiPrivateSwitchBase-root-79 MuiRadio-root-74 MuiRadio-colorPrimary-77 MuiPrivateSwitchBase-checked-80 MuiRadio-checked-75"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label-88"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-92"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
          </svg>
          <input
            checked={true}
            className="MuiPrivateSwitchBase-input-82"
            name="type"
            onChange={[Function]}
            type="radio"
            value="input"
          />
        </span>
        <span
          className="MuiTouchRipple-root-241"
        />
      </span>
      <span
        className="MuiTypography-root-101 MuiTypography-body1-110 MuiFormControlLabel-label-73"
      >
        Inputs
      </span>
    </label>
    <label
      className="MuiFormControlLabel-root-68"
    >
      <span
        className="MuiButtonBase-root-89 MuiIconButton-root-83 MuiPrivateSwitchBase-root-79 MuiRadio-root-74 MuiRadio-colorPrimary-77"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label-88"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-92"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
          </svg>
          <input
            checked={false}
            className="MuiPrivateSwitchBase-input-82"
            name="type"
            onChange={[Function]}
            type="radio"
            value="output"
          />
        </span>
        <span
          className="MuiTouchRipple-root-241"
        />
      </span>
      <span
        className="MuiTypography-root-101 MuiTypography-body1-110 MuiFormControlLabel-label-73"
      >
        Outputs
      </span>
    </label>
  </div>
  <div
    className="NeuronFilter-expandablePanel-137"
  >
    <div
      className="MuiPaper-root-145 MuiPaper-elevation1-148 MuiExpansionPanel-root-142"
    >
      <div
        aria-expanded={false}
        className="MuiButtonBase-root-89 MuiExpansionPanelSummary-root-172"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <div
          className="MuiExpansionPanelSummary-content-176"
        >
          <h6
            className="MuiTypography-root-101 MuiTypography-subtitle1-119"
          >
            Optional neuron/segment filters
          </h6>
        </div>
        <div
          aria-hidden="true"
          className="MuiButtonBase-root-89 MuiIconButton-root-83 MuiExpansionPanelSummary-expandIcon-177"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={-1}
        >
          <span
            className="MuiIconButton-label-88"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-92"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-241"
          />
        </div>
      </div>
      <div
        aria-hidden="true"
        className="MuiCollapse-container-178"
        style={
          Object {
            "minHeight": "0px",
          }
        }
      >
        <div
          className="MuiCollapse-wrapper-180"
        >
          <div
            className="MuiCollapse-wrapperInner-181"
          >
            <div
              className="MuiExpansionPanelDetails-root-182 NeuronFilter-nopad-138"
            >
              <div
                className="MuiFormControl-root-3 NeuronFilter-formControl-136"
              >
                <div
                  className="MuiFormControl-root-3 NeuronFilter-formControl-136"
                >
                  <label
                    className="MuiFormControlLabel-root-68"
                  >
                    <span
                      className="MuiSwitch-root-183"
                    >
                      <span
                        className="MuiButtonBase-root-89 MuiIconButton-root-83 MuiPrivateSwitchBase-root-79 MuiSwitch-switchBase-186 MuiSwitch-colorPrimary-188 MuiPrivateSwitchBase-checked-80 MuiSwitch-checked-187"
                        onBlur={[Function]}
                        onContextMenu={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={null}
                      >
                        <span
                          className="MuiIconButton-label-88"
                        >
                          <span
                            className="MuiSwitch-icon-184 MuiSwitch-iconChecked-185"
                          />
                          <input
                            checked={true}
                            className="MuiPrivateSwitchBase-input-82"
                            disabled={false}
                            onChange={[Function]}
                            type="checkbox"
                          />
                        </span>
                        <span
                          className="MuiTouchRipple-root-241"
                        />
                      </span>
                      <span
                        className="MuiSwitch-bar-191"
                      />
                    </span>
                    <span
                      className="MuiTypography-root-101 MuiTypography-body1-110 MuiFormControlLabel-label-73"
                    >
                      <h6
                        className="MuiTypography-root-101 MuiTypography-subtitle1-119"
                        style={
                          Object {
                            "display": "inline-flex",
                          }
                        }
                      >
                        Limit to neurons
                        <div
                          aria-describedby={null}
                          className="NeuronFilter-tooltip-141"
                          onBlur={[Function]}
                          onFocus={[Function]}
                          onMouseLeave={[Function]}
                          onMouseOver={[Function]}
                          onTouchEnd={[Function]}
                          onTouchStart={[Function]}
                          title="Limit to neurons (bodies with >=2 t-bars, >=10 psds, name, soma, or status)"
                        >
                          ?
                        </div>
                      </h6>
                    </span>
                  </label>
                </div>
                <div
                  className="MuiFormControl-root-3 MuiFormControl-marginDense-5 NeuronFilter-textField-140"
                >
                  <label
                    className="MuiFormLabel-root-18 MuiInputLabel-root-7 MuiInputLabel-formControl-12 MuiInputLabel-animated-15 MuiInputLabel-marginDense-13 MuiInputLabel-outlined-17"
                    data-shrink={false}
                  >
                    minimum # pre (optional)
                  </label>
                  <div
                    className="MuiInputBase-root-38 MuiOutlinedInput-root-200 MuiInputBase-formControl-39 MuiInputBase-marginDense-45"
                    onClick={[Function]}
                  >
                    <fieldset
                      aria-hidden={true}
                      className="MuiPrivateNotchedOutline-root-213 MuiOutlinedInput-notchedOutline-207"
                      style={
                        Object {
                          "paddingLeft": 8,
                        }
                      }
                    >
                      <legend
                        className="MuiPrivateNotchedOutline-legend-214"
                        style={
                          Object {
                            "width": 0.01,
                          }
                        }
                      >
                        <span
                          dangerouslySetInnerHTML={
                            Object {
                              "__html": "&#8203;",
                            }
                          }
                        />
                      </legend>
                    </fieldset>
                    <input
                      aria-invalid={false}
                      className="MuiInputBase-input-48 MuiOutlinedInput-input-208 MuiInputBase-inputType-51 MuiInputBase-inputMarginDense-49 MuiOutlinedInput-inputMarginDense-209"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      required={false}
                      rows={1}
                      type="number"
                      value=""
                    />
                  </div>
                </div>
                <div
                  className="MuiFormControl-root-3 MuiFormControl-marginDense-5 NeuronFilter-textField-140"
                >
                  <label
                    className="MuiFormLabel-root-18 MuiInputLabel-root-7 MuiInputLabel-formControl-12 MuiInputLabel-animated-15 MuiInputLabel-marginDense-13 MuiInputLabel-outlined-17"
                    data-shrink={false}
                  >
                    minimum # post (optional)
                  </label>
                  <div
                    className="MuiInputBase-root-38 MuiOutlinedInput-root-200 MuiInputBase-formControl-39 MuiInputBase-marginDense-45"
                    onClick={[Function]}
                  >
                    <fieldset
                      aria-hidden={true}
                      className="MuiPrivateNotchedOutline-root-213 MuiOutlinedInput-notchedOutline-207"
                      style={
                        Object {
                          "paddingLeft": 8,
                        }
                      }
                    >
                      <legend
                        className="MuiPrivateNotchedOutline-legend-214"
                        style={
                          Object {
                            "width": 0.01,
                          }
                        }
                      >
                        <span
                          dangerouslySetInnerHTML={
                            Object {
                              "__html": "&#8203;",
                            }
                          }
                        />
                      </legend>
                    </fieldset>
                    <input
                      aria-invalid={false}
                      className="MuiInputBase-input-48 MuiOutlinedInput-input-208 MuiInputBase-inputType-51 MuiInputBase-inputMarginDense-49 MuiOutlinedInput-inputMarginDense-209"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      required={false}
                      rows={1}
                      type="number"
                      value=""
                    />
                  </div>
                </div>
                <div
                  className="MuiFormControl-root-3 NeuronFilter-formControl-136"
                >
                  <label
                    className="MuiFormLabel-root-18"
                    style={
                      Object {
                        "display": "inline-flex",
                      }
                    }
                  >
                    Filter by status
                    <div
                      aria-describedby={null}
                      className="NeuronFilter-tooltip-141"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onMouseLeave={[Function]}
                      onMouseOver={[Function]}
                      onTouchEnd={[Function]}
                      onTouchStart={[Function]}
                      title=""
                    >
                      ?
                    </div>
                  </label>
                  <div
                    className="css-10nd86i NeuronFilter-select-139"
                    onKeyDown={[Function]}
                  >
                    <div
                      className="css-vj8t7z"
                      onMouseDown={[Function]}
                      onTouchEnd={[Function]}
                    >
                      <div
                        className="css-1hwfws3"
                      >
                        <div
                          className="css-1492t68"
                        >
                          Select...
                        </div>
                        <div
                          className="css-1g6gooi"
                        >
                          <div
                            className=""
                            style={
                              Object {
                                "display": "inline-block",
                              }
                            }
                          >
                            <input
                              aria-autocomplete="list"
                              autoCapitalize="none"
                              autoComplete="off"
                              autoCorrect="off"
                              disabled={false}
                              id="react-select-4-input"
                              onBlur={[Function]}
                              onChange={[Function]}
                              onFocus={[Function]}
                              spellCheck="false"
                              style={
                                Object {
                                  "background": 0,
                                  "border": 0,
                                  "boxSizing": "content-box",
                                  "color": "inherit",
                                  "fontSize": "inherit",
                                  "opacity": 1,
                                  "outline": 0,
                                  "padding": 0,
                                  "width": "1px",
                                }
                              }
                              tabIndex="0"
                              type="text"
                              value=""
                            />
                            <div
                              style={
                                Object {
                                  "height": 0,
                                  "left": 0,
                                  "overflow": "scroll",
                                  "position": "absolute",
                                  "top": 0,
                                  "visibility": "hidden",
                                  "whiteSpace": "pre",
                                }
                              }
                            >
                              
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        className="css-1wy0on6"
                      >
                        <span
                          className="css-d8oujb"
                        />
                        <div
                          aria-hidden="true"
                          className="css-1ep9fjw"
                          onMouseDown={[Function]}
                          onTouchEnd={[Function]}
                        >
                          <svg
                            aria-hidden="true"
                            className="css-19bqh2r"
                            focusable="false"
                            height={20}
                            viewBox="0 0 20 20"
                            width={20}
                          >
                            <path
                              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    className="MuiButtonBase-root-89 MuiButton-root-215 MuiButton-contained-226 MuiButton-containedPrimary-227 MuiButton-raised-229 MuiButton-raisedPrimary-230"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-216"
    >
      Submit
    </span>
    <span
      className="MuiTouchRipple-root-241"
    />
  </button>
</div>
`;
